src/plugins/data/common/index_patterns/fields/index_pattern_field.ts

```diff
@@ -0,0 +1,188 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import {
+  IFieldType,
+  KbnFieldType,
+  getKbnFieldType,
+  KBN_FIELD_TYPES,
+  FieldFormat,
+} from '../../../common';
+import { OnNotification, FieldSpec } from '../types';
+
+import { IndexPattern } from '../index_patterns';
+
+export class IndexPatternField implements IFieldType {
+  readonly spec: FieldSpec;
+  // not writable or serialized
+  readonly indexPattern: IndexPattern;
+  readonly displayName: string;
+  private readonly kbnFieldType: KbnFieldType;
+
+  constructor(
+    indexPattern: IndexPattern,
+    spec: FieldSpec,
+    displayName: string,
+    onNotification: OnNotification
+  ) {
+    this.indexPattern = indexPattern;
```

Comment:
i don't like the fact that we need to pass indexPattern in just to be able to get to formatter, could we:
- remove format function and have consumers always use `indexPattern.getFormatterForField(field)` ?
- or actually push that to the service, so we would do something like `indexPatterns.getFormatter(indexPattern, field)` and both IndexPattern and IndexPatternField get rid of that dependency (and can be static)

i am fine with doing this later, i am just wondering how reasonable you think that is ?

Reply:
100% This PR keeps the class interfaces largely the same. My next PR will address this.